src/coreclr/jit/gentree.cpp

```diff
@@ -30340,7 +30340,8 @@ bool GenTreeLclFld::IsOffsetMisaligned() const
 
 bool GenTree::IsInvariant() const
 {
-    return OperIsConst() || OperIs(GT_LCL_ADDR) || OperIs(GT_FTN_ADDR);
+    return OperIsConst() || OperIs(GT_LCL_ADDR) || OperIs(GT_FTN_ADDR) ||
+           (OperIs(GT_IND) && AsIndir()->IsInvariantLoad() && ((AsIndir()->gtFlags & GTF_SIDE_EFFECT) == 0));
```

Comment:
in what cases do we have a side-effect on an invariant load?

Reply:
There seem to be some invariant `IND`s that take `LCL_VAR` and have side effects that pass through here, not blocking them like this caused asserts further down the line.

Reply:
I don't think we should continue tolerating it

Reply:
Resolved as discussed on discord.